up-l1/README.adoc

```diff
@@ -28,179 +28,199 @@ SPDX-FileType: DOCUMENTATION
+[source]
+----
+onReceive(message: UMessage) : Void
+----
 
+==== Parameters
+.OnReceive Parameters
+[width="100%",cols="15%,15%,70%"]
 |===
+|Parameter | Type | Description
 
- * <<UStatus>> is defined in link:../basics/error_model.adoc[uProtocol Error Model]
-
-
-NOTE: uProtocol data model used by the uTransport interface are declared in https://github.com/eclipse-uprotocol/up-core-api[up-core-api]
-
-== UListener
-
-`UListener` is an interface used for receiving messages by the client uE from the uTransport layer. Listeners are used for receiving messages asynchronously by the uTransport layer.  
+| message
+| link:../basics/umessage.adoc[UMessage]
+| The newly received message.
 
-  * Type name *MUST* be `UListener`
-  * *MUST* contain the method `void onReceive(UMessage&)` that is called by the Transport to notify (callback) the client
+|===
 
 
 == UTransport
 
-The transport layer API responsible for point-2-point communication. The interface provides common functionality across all transport implementations.
-
-* *MUST* implement all APIs defined in this section
-
-=== Send()
+This is the main entry point into a transport's messaging functionality.
 
-Transmit a UMessage.  
-
-
-Communication protocols (a.k.a. uProtocol Transports) define their own Protocol Data Unit (PDU) that comprises of header and payload. Some transports header parameters map already to uProtocol UAttributes. Sending UAttributes in the transport's payload as well as in the header increases overhead for little to no gain. To address this issue,  <<Send()>> allows flexibility for uTransport implementers to define however they see fit what gets mapped into their transport header vs payload. 
+=== Send
 
-For example, if HartleyTransport can map only UPriority to its PDU header, then HartleyTransport would define a PDU payload type that includes UPayload and all the other UAttributes and send that new type in its PDU payload such that the receiver will not loose any metadata. 
+Clients use this method to transmit a single message.
 
-The API signature:
-
-`OUT UStatus send(UMessage&)`
+[source]
+----
+send(message: UMessage) : Future<Void>
+)
+----
 
-==== Parameters
 .Send Parameters
-[width="100%",cols="20%,80%"]
+[width="100%",cols="15%,25%,60%"]
 |===
-|Parameter | Description
-
-| UMessage
+|Parameter | Type | Description
 
-| UProtocol message containing the header (UAttributes) and payload (UPayload)
-
- * *MUST* be passed by reference
+| message
+| link:../basics/umessage.adoc[UMessage]
+| The message to send.
 
+| result
+| Future<Void>
+| A succeeded future if the message has been sent successfully. This means that the underlying messaging infrastructure has accepted the message for delivery. It does *not* (necessarily) mean that the message has reached its destination (yet). +
+Otherwise, a failed future which contains a `UCode` indicating the reason for the failure.
```

Comment:
Does this:
```
Otherwise, a failed future which contains a `UCode` indicating the reason for the failure.
```
mean that the [`send` signature](https://github.com/eclipse-uprotocol/up-spec/pull/134/files#diff-a034e4666a20b9cc4b66d777b10e9b4ffac86a7646a7dc79d13c7d067987436fR76) should read:
```
send(message: UMessage) : Future<Void, UCode>
```
?

Reply:
In Java, for example, the Future<T> class is parameterized with the type of successful completion only because the Future will also be failed with an Exception. Rust is the only language I know of where any type can be used to represent the erroneous outcome.
FMPOV it should be sufficient to define that the Future will be failed with _something_ that contains the relevant information (the UCode in this case). In the Java case this probably means that the Exception contains a UCode typed field. In Rust it might mean that the error type is UCode. Does that make sense?

Reply:
Thanks for the context. Makes sense.

Reply:
Maybe we should, instead of using the Future<T> return type, just declare the concrete return type (T) and then also declare an Exception being thrown. Would that help to make it easier to understand? We could then also add some general wording to a section called _Implementation Hints_ or similar, where we advise implementers to use non-blocking async calls. WDYT? @PLeVasseur @stevenhartley

Reply:
I like your idea of having a tips & tricks section for implementers in which to put some of this advice. I still think having some level of flexibility _even_ at uP-L1 for languages to write it the most reasonable way makes sense, which is where such a section could bring further value IMHO.